Port Shadowkin

Finket opened the pull request:
# Description

Port over Shadowkin. https://github.com/Simple-Station/Parkstation-Friendly-Chainsaw/issues/2
https://github.com/Simple-Station/Parkstation/pull/42 
This PR creates a lot of directories not yet ported to this rebase.
Many other directories which are required in the original PRs are not yet created / addressed.

---

# TODO

- [x] Fix empathy chat
- [ ] Make a base mob that isn't a respirator
- [x] Fix shadowkin needing to breath
- [x] Fix aghost can't see darkswapped shadowkin
- [x] Fix the weird spacing on spriteviews
- [x] Fix shadowkin can become psionic? (PotentialPsionic is inherited) ?? I don't know if PotentialPsionic "chance" works
- [x] Move over to NetEntity https://www.github.com/space-wizards/space-station-14/pull/18935 https://docs.spacestation14.com/en/robust-toolbox/netcode/net-entities.html
- [x] Adapt to the correct faction system
- [x] Action refactor https://www.github.com/space-wizards/space-station-14/pull/19900 https://www.github.com/space-wizards/space-station-14/discussions/19944
- [x] Minds relocation https://www.github.com/space-wizards/space-station-14/pull/19679
- [x] 42 https://github.com/Simple-Station/Parkstation/pull/42
- [x] 130 https://github.com/Simple-Station/Parkstation/pull/130
- [x] 146 https://github.com/Simple-Station/Parkstation/pull/146
- [x] 157 https://github.com/Simple-Station/Parkstation/pull/157
- [x] 188 https://github.com/Simple-Station/Parkstation/pull/188
- [x] 189 https://github.com/Simple-Station/Parkstation/pull/189
- [x] 218 https://github.com/Simple-Station/Parkstation/pull/218
- [ ] 209 https://github.com/Simple-Station/Parkstation/pull/209
- [ ] Other?

More todo, notes, and comments to address last (tentative):
- [x] Address removing SleepingComponent.WakeAction (in comments below)
- [x] What to do with VisibilityFlags? i.e. in Content.Server/...ShadowkinPowerSystem.DarkSwap.cs
- [x] Fix PlayerAttachedEvent in client systems
- [ ] Remaining Guidebook stuff should be ported separately
- [ ] https://github.com/space-wizards/space-station-14/pull/21276

I know this is a mess at the moment. Let me cook B-)

---

<details><summary><h1>Media</h1></summary>
<p>



</p>
</details>

---

DEATHB4DEFEAT commented:
:godmode:

Finket commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinPowerSystem.Rest.cs:
```diff
@@ -0,0 +1,79 @@
+            // Sleepy time
+            _entity.EnsureComponent<ForcedSleepingComponent>(args.Performer);
+            // No waking up normally (it would do nothing)
+            // Not sure what needs to be done here with the refactor
```
Not sure exactly how to get/remove this "Wake" action on the refactor

DEATHB4DEFEAT replied:
I think I just immediately removed the normal sleeping action, I don't know how actions work now

DEATHB4DEFEAT replied:
SleepingComponent has the wake action stored on it, Get the SleepingComponent (ForcedSleeping will add it) and remove SleepingComponent.WakeAction

Finket commented:
Couldn't keep looking at the "Entites" directory, had to fix the typo :D

DEATHB4DEFEAT commented:
> Couldn't keep looking at the "Entites" directory, had to fix the typo :D

Amazing

Finket commented:
Moved over to NetEntities, though my fixes are probably not ideal. The problem was mainly in the Events in shared.
Untested as well since other errors still remain.

Finket commented:
Code builds and runs now but:
- I get warnings for sprite issues and they don't all display correctly. Need to investigate which.
- Shadowkin actions don't work right now.

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinSystem.cs:
```diff
@@ -0,0 +1,236 @@
+                _power.TryBlackeye(uid);
+                Dirty(shadowkin);
+            }
+            // I can't figure out how to get this to go to the 100% filled state in the above if statement 😢
```
I don't think this is an issue still

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Speech/EntitySystems/ShadowkinAccentSystem.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        [Dependency] private readonly IRobustRandom _random = default!;
+
+        private static readonly Regex mRegex = new(@"[adgjmps]", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```
```suggestion
        private static readonly Regex mRegex = new(@"[adgjmpsv]", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Speech/EntitySystems/ShadowkinAccentSystem.cs:
```diff
@@ -0,0 +1,40 @@
+        }
+
+
+        public string Accentuate(string message)
+        {
+            message = message.Trim();
+
+            // Replace letters with other letters
+            message = mRegex.Replace(message, "m");
+            message = aRegex.Replace(message, "a");
+            message = rRegex.Replace(message, "r");
+
+            return message;
+        }
```
The first letter isn't capitalized when using this

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Species/Shadowkin/Components/ShadowkinComponent.cs:
```diff
@@ -0,0 +1,122 @@
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float MaxedPowerRoof = 0f;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float MaxedPowerRateMin = 45f;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float MaxedPowerRateMax = 150f;
```
This is too short

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Species/Shadowkin/Components/ShadowkinComponent.cs:
```diff
@@ -0,0 +1,122 @@
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float MinPowerRoof = 0f;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float MinPowerMin = 15f;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float MinPowerMax = 60f;
```
Decrease the max a bit, maybe 40?

```suggestion
    [ViewVariables(VVAccess.ReadWrite)]
    public float MinPowerMin = 15f;

    [ViewVariables(VVAccess.ReadWrite)]
    public float MinPowerMax = 40f;
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Species/Shadowkin/Components/ShadowkinComponent.cs:
```diff
@@ -0,0 +1,122 @@
+    public bool Blackeye = false;
+
+
+    // TODO This sucks, make this not hardcoded so we can do some fun things with it
```
Oh there's already a TODO here, amazing

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/Customization/tails.yml:
```diff
@@ -137,6 +137,15 @@
     - sprite: SimpleStation14/Mobs/Customization/tails32x32.rsi
       state: succubus
 
+- type: marking
+  id: TailTentacle
+  bodyPart: Tail
+  markingCategory: Tail
+  speciesRestriction: [nHuman, Oni]
+  sprites:
+    - sprite: SimpleStation14/Mobs/Customization/tails32x32.rsi
+      state: tentacle
+
```
```suggestion
```

DEATHB4DEFEAT commented on Resources/Prototypes/Species/human.yml:
```diff
@@ -58,6 +58,9 @@
     Arms:
       points: 4 # Parkstation-WideUseMarkings
       required: false
+    Wings:
+      points: 1
+      required: false
```
```suggestion
```

DEATHB4DEFEAT commented on Resources/Textures/SimpleStation14/Mobs/Customization/tails32x32.rsi/meta.json:
```diff
@@ -110,6 +110,10 @@
         {
             "name": "succubus",
             "directions": 4
+        },
+        {
+            "name": "tentacle",
+            "directions": 4
```
```suggestion
```

DEATHB4DEFEAT commented on Resources/Textures/SimpleStation14/Mobs/Customization/wings64x32.rsi/meta.json:
Delete this RSI
I can't comment on the file for some reason but remove the tentacle sprite too

DEATHB4DEFEAT commented on Resources/Textures/SimpleStation14/Mobs/Customization/wings96x34.rsi/meta.json:
Remove these additions

Finket commented:
Empathy and suffocation fixed. Major comments addressed.

Finket commented:
Fixed so that shadowkin effectively don't need to breath, but they are still BaseMobHuman and have the Respirator component. Also removed the psionic roll chance from the inherited PotentialPsionic component. As far as I can tell, Shadowkin should be in a totally usable state now.

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinPowerSystem.Darken.cs:
```diff
@@ -0,0 +1,144 @@
+    {
+        base.Update(frameTime);
+
+        // TODO Use a cached query, and update it on ShadowkinDarkSwappedComponent.Initialize
+        var shadowkins = _entity.EntityQueryEnumerator<ShadowkinDarkSwappedComponent>();
```
The darkened query should also be randomized like the light query

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinPowerSystem.Darken.cs:
```diff
@@ -0,0 +1,143 @@
+                // Not a light we should affect
+                if (!_entity.TryGetComponent(light, out ShadowkinLightComponent? shadowkinLight))
+                    continue;
+                // Not powered, undo changes
```
```suggestion

                // Not powered, undo changes
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinPowerSystem.Darken.cs:
```diff
@@ -0,0 +1,143 @@
+
+                // If the light isn't attached to an entity, attach it to this Shadowkin
+                if (shadowkinLight.AttachedEntity == EntityUid.Invalid)
+                {
+                    shadowkinLight.AttachedEntity = uid;
+                }
```
```suggestion
                    shadowkinLight.AttachedEntity = uid;

```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinPowerSystem.Darken.cs:
```diff
@@ -0,0 +1,143 @@
+                    shadowkin.DarkenedLights.Remove(light);
+                    continue;
+                }
+                // 3% chance to remove the attached entity so it can become another Shadowkin's light
```
```suggestion

                // 3% chance to remove the attached entity so it can become another Shadowkin's light
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinSystem.cs:
```diff
@@ -0,0 +1,236 @@
+        // Add/Remove the component, which should handle the rest
+        if (_entity.HasComponent<ShadowkinDarkSwappedComponent>(uid))
+            _entity.RemoveComponent<ShadowkinDarkSwappedComponent>(uid);
+        else
+            _entity.AddComponent<ShadowkinDarkSwappedComponent>(uid);
+    }
+
+    private void ForceTeleport(EntityUid uid, ShadowkinComponent component)
+    {
+        // Create the event we'll later raise, and set it to our Shadowkin.
+        var args = new ShadowkinTeleportEvent { Performer = uid };
+
+        // Pick a random location on the map until we find one that can be reached.
+        var coords = Transform(uid).Coordinates;
+        EntityCoordinates? target = null;
+
+        // It'll iterate up to 8 times, shrinking in distance each time, and if it doesn't find a valid location, it'll return.
+        for (var i = 8; i != 0; i--)
+        {
+            var angle = Angle.FromDegrees(_random.Next(360));
+            var offset = new Vector2((float) (i * Math.Cos(angle)), (float) (i * Math.Sin(angle)));
+
+            target = coords.Offset(offset);
+
+            if (_interaction.InRangeUnobstructed(uid, target.Value, 0,
+                    CollisionGroup.MobMask | CollisionGroup.MobLayer))
+                break;
+
+            target = null;
+        }
+
+        // If we didn't find a valid location, return.
+        if (target == null)
+            return;
+
+        args.Target = target.Value;
+
+        // Raise the event to teleport the Shadowkin.
+        RaiseLocalEvent(uid, args);
+    }
```
These should probably be a part of their power's systems

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Species/Shadowkin/Systems/ShadowkinPowerSystem.cs:
```diff
@@ -0,0 +1,277 @@
+    {
+        var ent = _entity.GetNetEntity(uid);
+        // Raise an attempted blackeye event
+        var ev = new ShadowkinBlackeyeAttemptEvent(ent);
+        RaiseLocalEvent(ev);
+        if (ev.Cancelled)
+            return false;
```
Ditto; by-ref?

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Magic/shadowkin.yml:
The value overrides on these power events should be moved to the defaults of those events for easier consistency

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Species/shadowkin.yml:
```diff
@@ -0,0 +1,161 @@
+  roundStart: true
+  prototype: MobShadowkin
+  sprites: MobShadowkinSprites
+  defaultSkinTone: "#FFFFFF"
```
Is there a default eye color or something similar? Would be good to avoid clueless people never getting powers.

DEATHB4DEFEAT replied:
Laki asked me about why their Shadowkin never got any powers and I realized that taking away their powers for simply having black eyes is pretty lame and is not communicated, I'll probably remove that and keep it a trait

Finket replied:
Would you like to do this on the next update or should I remove the eye color blackeye-ing here myself?

DEATHB4DEFEAT reviewed (changes requested):
Should I move the prebase Shadowkin update here or finish it there then redo it here?

DEATHB4DEFEAT commented:
After #56 gets merged the color tint shader needs to be toggleable via the client's DeltaV vision trait option in their settings, they should have a couple examples of how to do this.

VMSolidus commented on Resources/Prototypes/Parkstation/Entities/Mobs/Player/shadowkin.yml:
```diff
@@ -0,0 +1,144 @@
+      thresholds: # Weak
+        0: Alive
+        80: Critical
+        160: Dead
```
Should probably change Dead to 180, there's some issues with medical balance if the Crit-Dead range is not equal to 100.

VMSolidus commented on Resources/Prototypes/Parkstation/Traits/disabilities.yml:
```diff
@@ -9,3 +9,14 @@
+- type: trait
+  id: ShadowkinBlackeye
+  name: Blackeye
+  description: You lose your special Shadowkin powers, in return for some points.
+  cost: 5
+  category: Negative
+  whitelist:
+    species:
+      - Shadowkin
+  components:
+    - type: ShadowkinBlackeyeTrait
```
```suggestion
  description: You have lost your Shadowkin powers.
  whitelist:
    components:
      - Shadowkin
  components:
    - type: ShadowkinBlackeyeTrait
```

VMSolidus commented on Resources/Prototypes/Parkstation/Damage/modifier_sets.yml:
```diff
@@ -0,0 +1,13 @@
+    Piercing: 1.1
+    Asphyxiation: 0
+    Cold: 0.75
+    Heat: 1.2
```
```suggestion
    Heat: 1.5
```

VMSolidus commented on Resources/Prototypes/Parkstation/Entities/Mobs/Player/shadowkin.yml:
```diff
@@ -0,0 +1,144 @@
+      requiredLegs: 2
+    # - type: DiseaseCarrier
+      # diseaseResist: 0.1
+    - type: Flammable
+      damage:
+        types:
+          Heat: 1.5 # burn more
```
```suggestion
    - type: Flammable
```

DEATHB4DEFEAT commented:
Trystan has ported Shadowkin to EE, and I'll continue my development on them there unless opposed.

DEATHB4DEFEAT commented:
womp
